solana/indexer/damm_v2/indexer.go

```diff
@@ -0,0 +1,358 @@
+		}
+	}
+	return nil
+}
+
+func (d *Indexer) subscribe(ctx context.Context) ([]common.GrpcClient, error) {
+	done := false
+	page := 0
+	pageSize := MAX_POOLS_PER_SUBSCRIPTION
+	total := 0
+	grpcClients := make([]common.GrpcClient, 0)
+	for !done {
+		pools, err := getSubscribedDammV2Pools(ctx, d.pool, pageSize, page*pageSize)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get pools: %w", err)
+		}
+		if len(pools) == 0 {
+			d.logger.Info("no pools to subscribe to")
+			return grpcClients, nil
+		}
+		total += len(pools)
+
+		d.logger.Debug("subscribing to pools....", zap.Int("count", len(pools)))
+		subscription := d.makeSubscriptionRequest(ctx, pools)
+
+		// Handle each message from the subscription
+		handleMessage := func(ctx context.Context, msg *pb.SubscribeUpdate) {
+			err := d.HandleUpdate(ctx, msg)
+			if err != nil {
+				d.logger.Error("failed to handle update", zap.Error(err))
+
+				// Add messages that failed to process to the retry queue
+				if err := common.AddToRetryQueue(ctx, d.pool, NAME, msg, err.Error()); err != nil {
+					d.logger.Error("failed to add to retry queue", zap.Error(err))
+				}
+			}
+		}
+
+		grpcClient := d.grpcFactory(d.grpcConfig)
+		err = grpcClient.Subscribe(ctx, subscription, handleMessage, func(err error) {
+			d.logger.Error("error in subscription", zap.Error(err))
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to subscribe to pools: %w", err)
+		}
+		grpcClients = append(grpcClients, grpcClient)
+
+		if len(pools) < pageSize {
+			done = true
+		}
+		page++
+	}
+	d.logger.Info("subscribed to pools", zap.Int("count", total))
+	return grpcClients, nil
+}
+
+func (d *Indexer) makeSubscriptionRequest(ctx context.Context, dammV2Pools []string) *pb.SubscribeRequest {
+	commitment := pb.CommitmentLevel_CONFIRMED
+	subscription := &pb.SubscribeRequest{
+		Commitment: &commitment,
+	}
+
+	// Listen to all watched pools
+	subscription.Accounts = make(map[string]*pb.SubscribeRequestFilterAccounts)
+	accountFilter := pb.SubscribeRequestFilterAccounts{
+		Owner:   []string{meteora_damm_v2.ProgramID.String()},
+		Account: dammV2Pools,
+	}
+	subscription.Accounts[NAME] = &accountFilter
+
+	// Listen to all positions for each pool
+	for _, pool := range dammV2Pools {
+		accountFilter := pb.SubscribeRequestFilterAccounts{
+			Owner: []string{meteora_damm_v2.ProgramID.String()},
+			Filters: []*pb.SubscribeRequestFilterAccountsFilter{
+				{
+					Filter: &pb.SubscribeRequestFilterAccountsFilter_Memcmp{
+						Memcmp: &pb.SubscribeRequestFilterAccountsFilterMemcmp{
+							Offset: 8, // Offset of the pool field in the position account (after discriminator)
+							Data: &pb.SubscribeRequestFilterAccountsFilterMemcmp_Base58{
+								Base58: pool,
+							},
+						},
+					},
+				},
+				{
+					Filter: &pb.SubscribeRequestFilterAccountsFilter_Datasize{
+						Datasize: 408, // byte size of a Position account
```

Comment:
For some reason, a filter that is ostensibly "watch for accounts where the data is exactly 408 bytes and the field at byte offset 8 is this value" feels sketchy and fragile to me. I assume this is a common pattern for subscriptions? 😅 
They can't easily change this, right? Like the data size goes up by four bytes or something like that.